### DISCUSSION TITLE
Any plans of adding `if-else` chain + early-return support?

### DISCUSSION THREAD

**Discussion by rchougule** [antonmedv/expr#709] [Q&A]
Any plans of adding `if-else` chain with the option of returning a value for every condition, for e.g.:

```
let a;
let b;
if identifier == "X" {
  a = "SUCCESS"
  b = "REASON FOR SUCCESS"
} else if identifier2 == "Y" {
  a = "FAILED"
  b = "REASON FOR FAILURE"
} else {
  return {key: "value"}
}

<return> {result: a, reason: b}
```

Relying on ternary operator for this becomes slightly clumsy. Also, early returns when a check passes/fails would also help.

Thanks,
Rohan Chougule

**Comment by antonmedv** [comment]
I do have plans to add if-else support (as well as switch #586).

But I do plan to make if-else as expressions:

```f#
let foo = if cond1 {
  "success"
} else if cond2 {
  "alternative"
} else {
  "failure"
}
```

Even early returns:


```f#
if age < 3 {
  return false
}

if age < 10 {
  return false
}

true // User is mature enough to by a ticket.
```

But no variables reassignments.

**Reply by rchougule** [reply]
Got it. Thanks for the update.
Any idea what timeframe we are looking at?

**Reply by antonmedv** [reply]
I think within a month.

**Reply by rchougule** [reply]
> But no variables reassignments.

Is it possible to have reassignments along with multi-variable assignment at once?

@antonmedv Any reason for not doing this? We have use cases where we start with some default values for set of variables, and would want to update them later depending on how the expressions evaluate. For e.g.
```
let result = "NA";
let code = 0;
let reason = "";

result, reason = user.Id == 1 ? "PASSED", "reason x" : "FAILED", "reason y";
result, reason = result != "PASSED" && comment.Id == 1 ? "PASSED", "reason p" : "FAILED", "reason q";

code = result == "PASSED" ? 200 : 0;

{RESULT: result, CODE: code, REASON: reason}
```

re-assignment, along with doing it for multiple variables at once will help with writing more intuitive expressions.

Thanks again!

**Reply by antonmedv** [reply]
I see your point. Let me research this topic. We can probably figure out how to organically add variable reassignments.

**Reply by rchougule** [reply]
> I see your point. Let me research this topic. We can probably figure out how to organically add variable reassignments.

Thank you. Will revisit this thread after some time to check 😬.

Let me know if you'd need more inputs from me regarding this.

Thanks again.

**Comment by ryoheinagao** [comment]
Hi, I'm eager to use early-return feature. Since I wan to use some expensive process as function on Env. To reduce call of such functions, I'm glad to use early-return feature.

Is it still planned to be supported?

**Reply by ryoheinagao** [reply]
https://go.dev/play/p/SzFOuxVBplr might work for me. But it will bring many of if-else...

**Comment by ronaksahai** [comment]
Hey, is this still planned/in development?

**Reply by antonmedv** [reply]
Yes! else if is planned `if { ... } else if { ... } else { ... }` in v1.17.7.

For early return we need to discuss the language impact. Let's create a separate discussion for early return?

**Reply by tordrt** [reply]
Early return would be amazing. Nr 1 one my list. Would siginificantly reduce complexity for a lot of programs both to create them and also readability.  Currently there is no other way than doing a huge chain of conditionals on the end if a program starts getting a bit complex. Often some of these options could have been returned very early on in the program.
